### ISSUE TITLE
Test suite hangs on sparc because childs are not reaped

### ISSUE THREAD

**Issue by DerDakon** [fish-shell/fish-shell#6878]
Happens with both fish 3.0.2 and 3.1.

Downstream report: https://bugs.gentoo.org/709662

Log from 3.0.2:
```
#0  0x700fa9e0 in job_t::continue_job(bool) ()
#1  0x700aafb8 in exec_job(parser_t&, std::shared_ptr<job_t>) ()
#2  0x7012f448 in parse_execution_context_t::run_1_job(tnode_t<grammar::job>, block_t const*) ()
#3  0x7012fbe8 in parse_execution_context_t::run_job_conjunction(tnode_t<grammar::job_conjunction>, block_t const*) ()
#4  0x70133c0c in parse_execution_result_t parse_execution_context_t::run_job_list<grammar::job_list>(tnode_t<grammar::job_list>, block_t const*) ()
#5  0x7013118c in parse_execution_context_t::eval_node(tnode_t<grammar::job_list>, block_t const*, io_chain_t const&) ()
#6  0x700f0f5c in int parser_t::eval_node<grammar::job_list>(std::shared_ptr<parsed_source_t const>, tnode_t<grammar::job_list>, io_chain_t const&, block_type_t, std::shared_ptr<job_t>) ()
#7  0x700ee330 in parser_t::eval(std::shared_ptr<parsed_source_t const>, io_chain_t const&, block_type_t) ()
#8  0x700efe5c in parser_t::eval(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, io_chain_t const&, block_type_t) ()
#9  0x700acdf0 in exec_subshell_internal(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::vector<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::allocator<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > > >*, bool, bool) ()
#10 0x700b1d14 in expand_cmdsubst(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::vector<completion_t, std::allocator<completion_t> >*, std::vector<parse_error_t, std::allocator<parse_error_t> >*) ()
#11 0x700b25ec in expand_stage_cmdsubst(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::vector<completion_t, std::allocator<completion_t> >*, int, std::vector<parse_error_t, std::allocator<parse_error_t> >*) ()
#12 0x700b0570 in expand_string(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::vector<completion_t, std::allocator<completion_t> >*, int, std::vector<parse_error_t, std::allocator<parse_error_t> >*) ()
#13 0x7012b130 in parse_execution_context_t::expand_arguments_from_nodes(std::vector<tnode_t<grammar::argument>, std::allocator<tnode_t<grammar::argument> > > const&, std::vector<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::allocator<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > > >*, parse_execution_context_t::globspec_t) ()
#14 0x7012c974 in parse_execution_context_t::populate_plain_process(job_t*, process_t*, tnode_t<grammar::plain_statement>) ()
#15 0x7012d7b0 in parse_execution_context_t::populate_job_process(job_t*, process_t*, tnode_t<grammar::statement>) ()
#16 0x7012df4c in parse_execution_context_t::populate_job_from_job_node(job_t*, tnode_t<grammar::job>, block_t const*) ()
#17 0x7012f564 in parse_execution_context_t::run_1_job(tnode_t<grammar::job>, block_t const*) ()
#18 0x7012fbe8 in parse_execution_context_t::run_job_conjunction(tnode_t<grammar::job_conjunction>, block_t const*) ()
#19 0x70133c0c in parse_execution_result_t parse_execution_context_t::run_job_list<grammar::job_list>(tnode_t<grammar::job_list>, block_t const*) ()
#20 0x7013118c in parse_execution_context_t::eval_node(tnode_t<grammar::job_list>, block_t const*, io_chain_t const&) ()
#21 0x700f0f5c in int parser_t::eval_node<grammar::job_list>(std::shared_ptr<parsed_source_t const>, tnode_t<grammar::job_list>, io_chain_t const&, block_type_t, std::shared_ptr<job_t>) ()
#22 0x700ade34 in void internal_exec_helper<grammar::job_list>(parser_t&, std::shared_ptr<parsed_source_t const>, tnode_t<grammar::job_list>, io_chain_t const&, std::shared_ptr<job_t>) ()
#23 0x700ab510 in exec_job(parser_t&, std::shared_ptr<job_t>) ()
#24 0x7012f448 in parse_execution_context_t::run_1_job(tnode_t<grammar::job>, block_t const*) ()
#25 0x7012fbe8 in parse_execution_context_t::run_job_conjunction(tnode_t<grammar::job_conjunction>, block_t const*) ()
#26 0x70133c0c in parse_execution_result_t parse_execution_context_t::run_job_list<grammar::job_list>(tnode_t<grammar::job_list>, block_t const*) ()
#27 0x7013118c in parse_execution_context_t::eval_node(tnode_t<grammar::job_list>, block_t const*, io_chain_t const&) ()
#28 0x700f0f5c in int parser_t::eval_node<grammar::job_list>(std::shared_ptr<parsed_source_t const>, tnode_t<grammar::job_list>, io_chain_t const&, block_type_t, std::shared_ptr<job_t>) ()
#29 0x700ee330 in parser_t::eval(std::shared_ptr<parsed_source_t const>, io_chain_t const&, block_type_t) ()
#30 0x700efe5c in parser_t::eval(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, io_chain_t const&, block_type_t) ()
#31 0x700acdf0 in exec_subshell_internal(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::vector<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::allocator<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > > >*, bool, bool) ()
#32 0x700b1d14 in expand_cmdsubst(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::vector<completion_t, std::allocator<completion_t> >*, std::vector<parse_error_t, std::allocator<parse_error_t> >*) ()
#33 0x700b25ec in expand_stage_cmdsubst(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::vector<completion_t, std::allocator<completion_t> >*, int, std::vector<parse_error_t, std::allocator<parse_error_t> >*) ()
#34 0x700b0570 in expand_string(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::vector<completion_t, std::allocator<completion_t> >*, int, std::vector<parse_error_t, std::allocator<parse_error_t> >*) ()
#35 0x7012b130 in parse_execution_context_t::expand_arguments_from_nodes(std::vector<tnode_t<grammar::argument>, std::allocator<tnode_t<grammar::argument> > > const&, std::vector<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::allocator<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > > >*, parse_execution_context_t::globspec_t) ()
#36 0x7012c974 in parse_execution_context_t::populate_plain_process(job_t*, process_t*, tnode_t<grammar::plain_statement>) ()
#37 0x7012d7b0 in parse_execution_context_t::populate_job_process(job_t*, process_t*, tnode_t<grammar::statement>) ()
#38 0x7012df4c in parse_execution_context_t::populate_job_from_job_node(job_t*, tnode_t<grammar::job>, block_t const*) ()
#39 0x7012f564 in parse_execution_context_t::run_1_job(tnode_t<grammar::job>, block_t const*) ()
#40 0x7012fbe8 in parse_execution_context_t::run_job_conjunction(tnode_t<grammar::job_conjunction>, block_t const*) ()
#41 0x70133c0c in parse_execution_result_t parse_execution_context_t::run_job_list<grammar::job_list>(tnode_t<grammar::job_list>, block_t const*) ()
#42 0x7013118c in parse_execution_context_t::eval_node(tnode_t<grammar::job_list>, block_t const*, io_chain_t const&) ()
#43 0x700f0f5c in int parser_t::eval_node<grammar::job_list>(std::shared_ptr<parsed_source_t const>, tnode_t<grammar::job_list>, io_chain_t const&, block_type_t, std::shared_ptr<job_t>) ()
#44 0x700ee330 in parser_t::eval(std::shared_ptr<parsed_source_t const>, io_chain_t const&, block_type_t) ()
#45 0x700ff438 in read_ni(int, io_chain_t const&) ()
#46 0x70109c80 in reader_read(int, io_chain_t const&) ()
#47 0x70033af4 in main ()
```

What it does the whole time is just this:

```
_newselect(6, [5], NULL, NULL, {tv_sec=0, tv_usec=10000}) = 0 (Timeout)
_newselect(6, [5], NULL, NULL, {tv_sec=0, tv_usec=10000}) = 0 (Timeout)
```

This machine has many cores (64 threads total) which are rather slow on the other hand (1,4GHz), another machine (256 threads, 3,6GHz) shows the same behavior.

**Comment by ridiculousfish** [comment]
Neat! I wish I had a sparc machine to test with. I tried on a high-core x86 with gentoo and could not reproduce.

@DerDakon if you are willing, would you please try the following:

1. When fish is stuck like this, does sending SIGCHLD to fish unstick it? for example `kill -CHLD <pid>`

2. In the backtrace, are there other threads running? What is their backtrace?

Thank you for filing!

**Comment by DerDakon** [comment]
Sending SIGCHLD does not help. There is only one thread.

**Comment by ridiculousfish** [comment]
I have been trying to get a Debian or Gentoo SPARC64 system running in qemu, without success - both are apparently able to install, but fail early boot in weird ways.

If someone has a suggestion for how to get access to a SPARC64 system, even in emulation, please share.

**Comment by mattst88** [comment]
> If someone has a suggestion for how to get access to a SPARC64 system, even in emulation, please share.

Please email an ssh key and your desired username to my username at gentoo.org and I'll create an account for you on our SPARC64 system.

**Comment by ridiculousfish** [comment]
Through mattst88's very kind offer I was able to test on the SPARC64 system, where the bug reproduces easily.

**tl;dr `posix_spawn` is invoking the `rt_sigprocmask` syscall with bogus params and not checking the result. I suspect the problem is gcc has miscompiled glibc.**

The high-level problem is that `posix_spawn` is leaving the proc's sigmask set to an uninitialized variable. [Reduced test case here](https://gist.github.com/ridiculousfish/1b28af43ae2a3a871a5f4aa43a3811f2). Expected output:

```
> gcc test.c && ./a.out
sigmask before posix_spawn
  (empty)
sigmask after posix_spawn
  (empty)
hello
```

but on the SPARC system we see `posix_spawn` leaves the sigmask set with random bits.

Through `strace` on this file we can observe the syscall traffic:

```
mmap(NULL, 40960, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_STACK, -1, 0) = 0xfff8000100022000
rt_sigprocmask(0xff87ea08 /* SIG_??? */, ~[], 0x7feff87ea88, 8) = -1 EINVAL (Invalid argument)
clone(child_stack=0xfff800010002b741, flags=CLONE_VM|CLONE_VFORK|SIGCHLD) = 2852627
munmap(0xfff8000100022000, 40960)       = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
```

Notice the first `rt_sigprocmask` call is failing with `EINVAL` because the first argument is a random pointer (should be 1, which is `SIG_BLOCK`). So the current sigmask is not successfully saved, so when it is restored, it gets set to an uninitialized value. Shame on `posix_spawn` for not performing error checking here.

I pulled the glibc sources from `/var/db/repos/gentoo/distfiles/glibc-2.29.tar.xz`. [Here is the failing call](https://gist.github.com/ridiculousfish/0193393028a3075252b298376251c2ed#file-spawni-c-L372).

Disassembling this code, I see (note AT&T syntax, so `mov %A %B` copies A to B):

   0xfff8000100205f44:	mov  %o0, %o1     <-- pointer to filled sigset
   0xfff8000100205f48:	mov  %i3, %o2     <-- pointer to storage for saved set
   0xfff8000100205f4c:	mov  8, %o3       <-- size of sigset, which is 8 bytes (64 signals)
   0xfff8000100205f50:	mov  0x67, %g1    <-- syscall number?
   0xfff8000100205f54:	ta  0x6d          <-- trap
   
what's missing is setting `%o0` to 1 (`SIG_BLOCK`). Perhaps this instruction was dropped when `__libc_signal_block_all` was inlined. This is why I suspect gcc.

This is as far as I can get without recompiling glibc, but I am not sure how to do that on gentoo.

@mattst88 there is a serious bug which makes `posix_spawn` unusable on gentoo/SPARC64; [here's a simple test case](https://gist.github.com/ridiculousfish/1b28af43ae2a3a871a5f4aa43a3811f2). I think the next step would be to add error checking to the  `__libc_signal_block_all` call shown above, and then reduce `__spawnix` to produce a reduced test case for gcc (perhaps in the process discovering it is glibc).

**Comment by ridiculousfish** [comment]
I should note how to work around this in fish by disabling its use of `posix_spawn`:

1. At build time by defining `FISH_USE_POSIX_SPAWN` to 0.
2. Or, at runtime by setting the `fish_use_posix_spawn` environment variable to 0.

**Comment by trofi** [comment]
`rt_sigprocmask` rings a bell: https://sourceware.org/git/?p=glibc.git;a=commitdiff;h=b33e946fbb1659d2c5937c4dd756a7c49a132dff

`posix_spawn` might also be related to register clobbering: https://sourceware.org/git/?p=glibc.git;a=commitdiff;h=c261718b3fda4b629ee120f48e3ee50db9c793a1

Worth double-checking if we have these patches on tested glibc.

**Comment by trofi** [comment]
> `rt_sigprocmask` rings a bell: https://sourceware.org/git/?p=glibc.git;a=commitdiff;h=b33e946fbb1659d2c5937c4dd756a7c49a132dff
> 
> `posix_spawn` might also be related to register clobbering: https://sourceware.org/git/?p=glibc.git;a=commitdiff;h=c261718b3fda4b629ee120f48e3ee50db9c793a1
> 
> Worth double-checking if we have these patches on tested glibc.

`rt_sigprocmask` is present in both `2.30-r8` and `2.31-r3`. `posix_spawn` is not present in either of those.

**Comment by trofi** [comment]
Surprisingly the bug also manifests on sparc32 userland. Looking at https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/spawni.c;h=f157bfffd2e480f1ffb667714800ccbb838ec7c8;hb=HEAD#l372

```c
 349   ...
 350   void *stack = __mmap (NULL, stack_size, prot,
 351                         MAP_PRIVATE | MAP_ANONYMOUS | MAP_STACK, -1, 0);
 ...
 372   __libc_signal_block_all (&args.oldmask);
 ...
```

`__libc_signal_block_all` is an asm wrapper: https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/internal-signals.h;h=3fbd4807d117337ddf82ab18617e97eb4e4e7928;hb=HEAD#l66

From your analysis we break very first literal parameter of `rt_sigprocmask` (`internal_syscall4` wrapper). 

I'll test if register clobbering patch fixes it and report back. If it still does not work we can tweak the wrapper further to stop claiming register outside asm inline block.

**Comment by DerDakon** [comment]
Using 2.31-r3 I see the same errors in fkr.expect with and without the patches you attached to the Gentoo bug when running the testsuite of 3.1.2. The hangs are gone in both.

**Comment by trofi** [comment]
Backported register clobber to `glibc-2.30` and `glibc-2.31` as https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=97e3d34a588fad1db19369bbe9787d1c21fc3616

At least on `glibc-2.30` it makes `posix_spawn` test pass without erroneous signal blockers on sparc32 userland. `glibc-2.31` seems to be different enough not to exhibit exactly the same failure without the patch. But it was probably broken in other obscure ways.

**Comment by faho** [comment]
It seems like this is a glibc bug and there's nothing to do in fish. Closing.
